### PR TITLE
Check the presence of glyphs when selecting font

### PR DESCRIPTION
**Pull request by bwikbs**
PR due to default branch change
 - previous PR : https://github.com/flutter-tizen/engine/pull/106
----
This PR is to solve the font issue by re-enable ``font config`` and changing font selection logic(with workaround)
PR that uses the font config as the default and reverts our modification will be prepared separately.
Also, it contains the following minor change.
- Change fallback font name to ``TizenDefaultFont`` :  The font config setting in the Tizen device has an alias called ``TizenDefaultFont``, and it is connected to the actual font in device. So if we decide to use the font config we can use it as a fallback font family.

### CODE REVIEW

**Comment by bwikbs** [comment]
After uploading to 2.2.1, a new problem occurred. :cry: 
dump_screen

**Review by bbrto21 — APPROVED** [review]
The workaround added to the sorting algorithm seems to be the best we can do right now as long as we should use SamsungOneUI font family

File: third_party/txt/src/minikin/FontFamily.cpp [hunk 1]
```diff
@@ -146,15 +146,28 @@ static FontFakery computeFakery(FontStyle wanted, FontStyle actual) {
...
 }
 
-FakedFont FontFamily::getClosestMatch(FontStyle style) const {
+FakedFont FontFamily::getClosestMatch(
+    FontStyle style,
+    uint32_t codepoint /* = 0 */,
+    uint32_t variationSelector /* = 0 */) const {
+  int bestMatch = INT_MAX;
   const Font* bestFont = nullptr;
-  int bestMatch = 0;
```
Comment (swift-kim) [inline third_party/txt/src/minikin/FontFamily.cpp:151]: It would be better to keep the variable order as it has been.

File: third_party/txt/src/minikin/FontFamily.cpp [hunk 2]
```diff
@@ -146,15 +146,28 @@ static FontFakery computeFakery(FontStyle wanted, FontStyle actual) {
...
   for (size_t i = 0; i < mFonts.size(); i++) {
     const Font& font = mFonts[i];
     int match = computeMatch(font.style, style);
-    if (i == 0 || match < bestMatch) {
-      bestFont = &font;
-      bestMatch = match;
+    bool result = false;
+    if (codepoint != 0) {
+      hb_font_t* hb_font = getHbFontLocked(font.typeface.get());
```
Comment (swift-kim) [inline third_party/txt/src/minikin/FontFamily.cpp:160]: nit:
```suggestion
      hb_font_t* hbFont = getHbFontLocked(font.typeface.get());
```

File: third_party/txt/src/minikin/FontFamily.cpp [hunk 3]
```diff
@@ -146,15 +146,28 @@ static FontFakery computeFakery(FontStyle wanted, FontStyle actual) {
...
+      uint32_t unusedGlyph = 0;
+      result = hb_font_get_glyph(hb_font, codepoint, variationSelector,
+                                 &unusedGlyph);
+      hb_font_destroy(hb_font);
+    }
+    if (!codepoint || (codepoint && result)) {
```
Comment (swift-kim) [inline third_party/txt/src/minikin/FontFamily.cpp:166]:
```suggestion
    if (codepoint == 0 || (codepoint != 0 && result)) {
```
For consistency.

File: third_party/txt/src/txt/font_collection.cc [hunk 4]
```diff
@@ -250,6 +250,28 @@ void FontCollection::SortSkTypefaces(
         int a_delta = std::abs(a_style.width() - SkFontStyle::kNormal_Width);
         int b_delta = std::abs(b_style.width() - SkFontStyle::kNormal_Width);
 
+        {
+          // A workaround to prevent emoji fonts being selected for normal text
+          // when normal and emojis are mixed at same font family.
```
Comment (swift-kim) [inline third_party/txt/src/txt/font_collection.cc:255]: You may move this code block before line 247.

File: third_party/txt/src/txt/font_collection.cc [hunk 5]
```diff
@@ -250,6 +250,28 @@ void FontCollection::SortSkTypefaces(
...
+
+          bool a_isContainEmoji = false;
+          bool b_isContainEmoji = false;
```
Comment (swift-kim) [inline third_party/txt/src/txt/font_collection.cc:258]: From what I guess from the code the correct naming would be
```suggestion
          bool a_isEmojiFont = false;
          bool b_isEmojiFont = false;
```

Otherwise it could also be
```suggestion
          bool a_containsEmoji = false;
          bool b_containsEmoji = false;
```

File: third_party/txt/src/txt/platform_linux.cc [hunk 6]
```diff
@@ -13,6 +13,9 @@
 namespace txt {
 
 std::vector<std::string> GetDefaultFontFamilies() {
+#ifdef FLUTTER_USE_FONTCONFIG
+  return {"TizenDefaultFont"};
+#else
```
Comment (swift-kim) [inline third_party/txt/src/txt/platform_linux.cc:18]: Do we need this change (caf01dd6f3422bbdd5b4dae3a441f0d4401caad5) if we're not to re-enable fontconfig?

**Review by swift-kim — COMMENTED** [review]
Are you going to request this change for merge (maybe partially) in the upstream?

**Comment by bwikbs** [comment]
> Are you going to request this change for merge (maybe partially) in the upstream?

@swift-kim I think I can do some of it ( yes, maybe.. I didn't really think about how far it would go.), but overall no. Most of the problems come from the unusual setting of product.

File: third_party/txt/src/minikin/FontCollection.cpp [hunk 7]
```diff
@@ -528,8 +528,27 @@ void FontCollection::itemize(const uint16_t* string,
...
+        size_t start = utf16Pos;
+        auto current_font = family->getClosestMatch(style, ch, variant);
+        if (result->back().fakedFont.font != current_font.font) {
+          const size_t prevChLength = U16_LENGTH(prevCh);
+          run->end -= prevChLength;
+          if (run->start == run->end) {
+            result->pop_back();
+          }
+          start -= prevChLength;
+        }
```
[the remark this answers is not in the document]
Comment (bwikbs) [reply third_party/txt/src/minikin/FontCollection.cpp:550]: This is the logic taken from the logic that handles the emoji above. 
What we are doing here is to check ,ultimately,  if we can draw a previous glyph using the font that draws the current glyph if certain conditions are met.

File: third_party/txt/src/txt/font_collection.cc (hunk 4 above)
Comment (bwikbs) [reply third_party/txt/src/txt/font_collection.cc:255]: Is line 247 the original part?

File: third_party/txt/src/txt/platform_linux.cc (hunk 6 above)
Comment (bwikbs) [reply third_party/txt/src/txt/platform_linux.cc:18]: No, ``TizenDefaultFont`` is not an actual font. That's an alias in fontconfig

File: third_party/txt/src/txt/font_collection.cc (hunk 4 above)
Comment (swift-kim) [reply third_party/txt/src/txt/font_collection.cc:255]: Before the declaration of `a_style`, because none of `a_style`, ..., `b_delta` is used in this block.

File: third_party/txt/src/txt/platform_linux.cc (hunk 6 above)
Comment (swift-kim) [reply third_party/txt/src/txt/platform_linux.cc:18]: Then you might revert the commit.

File: third_party/txt/src/minikin/FontCollection.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (swift-kim) [reply third_party/txt/src/minikin/FontCollection.cpp]: From what I can see in the comment for the original code that you copied from, the if block is supposed to be a workaround for a combined mark+emoji modifier character. (Here is its use case: https://emojipedia.org/emoji-modifier-sequence, just in case.) Are you sure the code acts as you expected? Wouldn't the rest of the code without the if block just do?

> if we can draw a previous glyph using the font that draws the current glyph

If so, how about a previous-previous glyph? Are you taking account of various cases?

[the remark this answers is not in the document]
Comment (bbrto21) [reply third_party/txt/src/minikin/FontCollection.cpp]: @swift-kim 
I think... this was added because of my misunderstanding about [this](https://github.com/flutter-tizen/engine/pull/106#issuecomment-859200687).
I guessed there is a problem that the character styles of sentences specified in one style seem to be different when too many fonts are unintentionally selected even though they belong to the same font family. But this seems to be all wrong.

Basically, if some fonts belongs to the same font family, there should be no awkwardness when displayed. I think this problem is also due to the font for the emoji too, So now I think that this is an unnecessary if statement

@bwikbs I've already checked about the above, but could you please check it again and update the PR?

File: third_party/txt/src/txt/font_collection.cc (hunk 4 above)
Comment (bwikbs) [reply third_party/txt/src/txt/font_collection.cc:255]: Oh.. It turns out I didn't understand your intent. I thought you were talking about the comment section only!

File: third_party/txt/src/minikin/FontCollection.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (bwikbs) [reply third_party/txt/src/minikin/FontCollection.cpp]: @swift-kim  Hmm... I'll update this part after I talk to @bbrto21 to figure out my understanding is correct.
Anyway, thanks for the thoughtful review!

File: third_party/txt/src/txt/platform_linux.cc (hunk 6 above)
Comment (bwikbs) [reply third_party/txt/src/txt/platform_linux.cc:18]: I'm sorry but I don't get the point. Because..?

Comment (swift-kim) [reply third_party/txt/src/txt/platform_linux.cc:18]: **The** commit in this context is caf01dd6f3422bbdd5b4dae3a441f0d4401caad5.

File: third_party/txt/src/minikin/FontCollection.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (bwikbs) [reply third_party/txt/src/minikin/FontCollection.cpp]: > Are you sure the code acts as you expected? 

Yes, it works exactly as intended. However, I agree that this solution itself is a workaround for our specific situation and is no longer needed(As @bbrto21  pointed out...).

> If so, how about a previous-previous glyph? Are you taking account of various cases?

You pointed out a weakness of this workaround! :smile:  At first I wanted to check the whole string, but it seemed too complicated and would cause performance issue. And I thought that our current situation is a font config setting problem, I am going to end it at this level. (Yes, this is an excuse for why this workaround came out.)

Anyway, I will remove this block and move the condition to the above.

[the remark this answers is not in the document]
Comment (bwikbs) [reply third_party/txt/src/minikin/FontCollection.cpp]: @swift-kim @bbrto21 I've update code! How about now?

[the remark this answers is not in the document]
Comment (swift-kim) [reply third_party/txt/src/minikin/FontCollection.cpp]: Looks good to me!

File: third_party/txt/src/txt/platform_linux.cc (hunk 6 above)
Comment (bwikbs) [reply third_party/txt/src/txt/platform_linux.cc:18]: I'm still not sure if I understood your intentions... 

This PR is base on ``fontconfig``.  And PR that uses font config again and revert unnecessary our modification will be prepared another PR (Perhaps @bbrto21 Am I right boram? :smile: ).
The reason for doing that is because it would be good to track the related ``fontconfig`` in the future.(Just my opinion)

I'm sorry for the lack of explanation of this.

Comment (swift-kim) [reply third_party/txt/src/txt/platform_linux.cc:18]: https://github.com/flutter-tizen/engine/pull/106#pullrequestreview-681361371 by @bbrto21:

> I think you need to re-enable fontconfig in azure-pipelines.

https://github.com/flutter-tizen/engine/pull/106#issuecomment-859266980 by @bwikbs:

> Reverting existing code and applying fontconfig seem to be a little different topic, so I would like to talk about it in another PR.

If this PR depends on fontconfig, why not re-enable the fontconfig option in azure-pipelines.yml? The change inside the `#ifdef FLUTTER_USE_FONTCONFIG` block doesn't affect anything without the option enabled, so I thought you could safely revert the change. I'm sorry but I still don't get your point.

Comment (bwikbs) [reply third_party/txt/src/txt/platform_linux.cc:18]: @swift-kim Well... You're right. I just wanted to collect the related re-enable font config contents in one PR. 
There is no great reason...(To be honest, I didn't want to revert and clean-up the existing ones. And because it seems that @bbrto21 has already done that work, I was going to ask for him)
I didn't think much about it.. If it's too weird, I'll enable the font config right away.

Comment (swift-kim) [reply third_party/txt/src/txt/platform_linux.cc:18]: Okay, if there will be a follow-up PR, you may just briefly mention what changes will be made in that PR and how `TizenDefaultFont` is related to those changes.

Comment (bwikbs) [reply third_party/txt/src/txt/platform_linux.cc:18]: I've updated PR description!
